File: javascript/57.insert-interval.js
/*
 * @lc app=leetcode id=57 lang=javascript
 *
 * [57] Insert Interval
 */

// @lc code=start
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function (intervals, newInterval) {
  if (
    intervals.length === 0 ||
    intervals[intervals.length - 1][1] < newInterval[0]
  ) {
    intervals.push(newInterval);
  } else {
    let l = Math.max(
      intervals.findIndex((interval) => interval[1] >= newInterval[0]),
      0
    );
    let r = l;
    for (; r < intervals.length; r++) {
      if (intervals[r][0] > newInterval[1]) break;
    }
    let slice = intervals.slice(l, r);
    if (slice.length) {
      newInterval[0] = Math.min(newInterval[0], slice[0][0]);
      newInterval[1] = Math.max(newInterval[1], slice[slice.length - 1][1]);
    }
    intervals.splice(l, slice.length, newInterval);
  }
  return intervals;
};
// @lc code=end

File: rust/57.insert-interval.rs
/*
 * @lc app=leetcode id=57 lang=rust
 *
 * [57] Insert Interval
 */

// @lc code=start
impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        if intervals.len() == 0 || intervals[intervals.len() - 1][1] < new_interval[0] {
            intervals.push(new_interval);
            return intervals;
        }
        let mut new_interval = new_interval;
        let l = intervals.partition_point(|interval| interval[1] < new_interval[0]);
        let r = intervals[l..].partition_point(|interval| interval[0] <= new_interval[1]) + l;
        let slice: Vec<Vec<i32>> = intervals.drain(l..r).collect();
        if !slice.is_empty() {
            new_interval[0] = new_interval[0].min(slice[0][0]);
            new_interval[1] = new_interval[1].max(slice[slice.len() - 1][1]);
        }
        intervals.insert(l, new_interval);
        intervals
    }
}
// @lc code=end
